examples/hitl/rearrange_v2/rearrange_v2.py

```diff
@@ -306,31 +284,62 @@ def __init__(self, app_service: AppService):
                     gui_agent_controller=self._gui_agent_controllers[
                         user_index
                     ],
-                    client_helper=self._client_helper,
                     server_sps_tracker=self._sps_tracker,
                 )
             )
 
+        # Reset the environment immediately.
+        self.on_environment_reset(None)
+
+    def get_next_state(self) -> Optional[AppStateBase]:
+        # If cancelled, skip upload and clean-up.
+        if self._cancel or self._is_episode_finished():
+            return create_app_state_reset(self._app_service, self._app_data)
+        else:
+            return None
+
+    def on_enter(self):
+        super().on_enter()
+
+        user_index_to_agent_index_map: Dict[int, int] = {}
+        for user_index in range(len(self._user_data)):
+            user_index_to_agent_index_map[user_index] = self._user_data[
+                user_index
+            ].gui_agent_controller._agent_idx
+
+    def on_exit(self):
+        super().on_exit()
+
+    def _is_episode_finished(self) -> bool:
+        """
+        Determines whether all users have finished their tasks.
+        """
+        return all(
+            self._user_data[user_index].episode_finished
+            for user_index in self._users.indices(Mask.ALL)
+        )
+
     def on_environment_reset(self, episode_recorder_dict):
         self._world.reset()
 
+        # Reset AFK timers.
+        # TODO: Move to idle_kick_timer class. Make it per-user. Couple it with "user_data" class
+        # TODO
+        self._app_service.remote_client_state._client_helper.activate_users()
+
         # Set the task instruction
         current_episode = self._app_service.env.current_episode
+        if hasattr(current_episode, "instruction"):
+            task_instruction = current_episode.instruction
+            # TODO: Users will have different instructions.
+            for user_index in self._users.indices(Mask.ALL):
+                self._user_data[user_index].task_instruction = task_instruction
```

Comment:
Hmmm, I see. I am setting such metadata/data for llm-agent directly from habitat `Env` class. It'd be good to standardize all data-setting to flow from here.


Therefore, interested to see how can I consume this `user-data` variable at init/reset time. We can just pass this on to `ctrl_helper`, right? It'll just be ignore as part of `*args, **kwargs` for other controllers which have no use for it?

Reply:
That should be doable.

If you only need user events, you should be able to bind events from `rearrange_v2.on_environment_reset()` as you will have access to both variables. I'll be adding them in the next 2 PRs.